### PR TITLE
feat(ingest/powerbi): powerbi dataset profiling

### PR DESCRIPTION
**Pull request by looppi**
DAX REST API based profiling implementation to query some details of PowerBI Datasets in the workspace. Does not introduce any breaking changes.

This implementation counts only amount of distinct values, min and max of a column. So it won't completely fulfill the possibilities Datahub UI provides. Extending the implementation to include other metrics isn't very complicated.

Is this something you would like to accept and include in Datahub or is using the DAX API for gathering information considered not a good practice?

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py [hunk 1]
```diff
@@ -47,6 +50,13 @@ def __init__(self, config: PowerBiDashboardSourceConfig) -> None:
             tenant_id=self.__config.tenant_id,
         )
 
+        self.__profiling_resolver = PowerBiDatasetProfilingResolver(
```
Comment (siddiquebagwan) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py:53]: There are only two way to get the metadata from PowerBI either using regular API or Admin API. You can clearly see this segregation in PowerBI Rest API documentation: https://learn.microsoft.com/en-us/rest/api/power-bi/.

As per doc it looks like it comes under regular-api, so please add profile_dataset in base resolver  and provide implementation in both regular and admin resolver (here it is just pass). Invoke the  powerbi_profiler.py from regular api's profile_dataset  method

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_classes.py [hunk 2]
```diff
@@ -108,6 +112,10 @@ class Measure:
         BooleanTypeClass, DateTypeClass, NullTypeClass, NumberTypeClass, StringTypeClass
     ] = dataclasses.field(default_factory=NullTypeClass)
     description: Optional[str] = None
+    min: Optional[str] = None
```
Comment (siddiquebagwan) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_classes.py:115]: Better to have MeasureProfiling dataclass and add the reference here

File: metadata-ingestion/docs/sources/powerbi/powerbi_pre.md [hunk 3]
```diff
@@ -107,6 +107,10 @@ By default, extracting endorsement information to tags is disabled. The feature
 
 Please note that the default implementation overwrites tags for the ingested entities, if you need to preserve existing tags, consider using a [transformer](../../../../metadata-ingestion/docs/transformer/dataset_transformer.md#simple-add-dataset-globaltags) with `semantics: PATCH` tags instead of `OVERWRITE`.
 
+## Profiling
+
+The profiling implementation is done through querying [DAX query endpoint](https://learn.microsoft.com/en-us/rest/api/power-bi/datasets/execute-queries). Therefore the principal needs to have permission to query the datasets to be profiled. Profiling is done with column based queries to be able to handle wide datasets without timeouts.
```
Comment (siddiquebagwan) [inline metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:112]: Additional documentation need to be added: 

1. Please also check this doc: https://datahubproject.io/docs/quick-ingestion-guides/powerbi/setup and add/update any steps if any specific permission is required for profiling. The markdown file is available at `docs/quick-ingestion-guides/powerbi/setup.md`
2. On Source class add decorator `@capability(
    SourceCapability.DATA_PROFILING,
    "Optionally enabled via configuration `profiling.enabled`",
)`

**Review by siddiquebagwan — CHANGES_REQUESTED** [review]

**Comment by looppi** [comment]
Thank you for the feedback! I'll start working with this asap.

**Comment by hsheth2** [comment]
@sid-acryl can you take another look at this PR?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/profiling_utils.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import re
+from typing import Dict, List
+
+
+def get_column_name(table_and_col: str) -> str:
```
Comment (sid-acryl) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/profiling_utils.py:5]: Make it -> Optional[str]

**Review by sid-acryl — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py [hunk 5]
```diff
@@ -515,6 +535,131 @@ def _get_pages_by_report(self, workspace: Workspace, report_id: str) -> List[Pag
...
+            headers=self.get_authorization_header(),
+        )
+        response.raise_for_status()
+        return response.json()
+
+    def _get_row_count(self, dataset: PowerBIDataset, table: Table) -> int:
+        query = DaxQuery.row_count_query(table.name)
+        try:
+            data = self._execute_profiling_query(dataset, query)
+            rows = data["results"][0]["tables"][0]["rows"]
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py:572]: it looks like they're already catching KeyError below, so it should be ok

**Comment by looppi** [comment]
Thank you for the review and feedback! I'll continue working on this still this week.

File: metadata-ingestion/docs/sources/powerbi/powerbi_pre.md [hunk 6]
```diff
@@ -107,6 +107,14 @@ By default, extracting endorsement information to tags is disabled. The feature
...
+## Profiling
+
+The profiling implementation is done through querying [DAX query endpoint](https://learn.microsoft.com/en-us/rest/api/power-bi/datasets/execute-queries). Therefore the principal needs to have permission to query the datasets to be profiled. Usually this means that the service principal should have `Contributor` role for the workspace to be ingested. Profiling is done with column based queries to be able to handle wide datasets without timeouts.
+
+Take into account that the profiling implementation exeutes fairly big amount of DAX queries and for big datasets this is substantial load to the PowerBI system.
+
+The `profiling_pattern` setting may be used to limit profiling actions to only a certain set of resources in PowerBI. Both allow and deny rules are matched against following pattern for every table in a PowerBI Dataset: `workspace_name.dataset_name.table_name`. User may limit profiling with these settings at table level, dataset level or workspace level.
+
 ## Admin Ingestion vs. Basic Ingestion
 PowerBI provides two sets of API i.e. [Basic API and Admin API](https://learn.microsoft.com/en-us/rest/api/power-bi/).
```
[the remark this answers is not in the document]
Comment (looppi) [reply metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:120]: Sorry for the delay, I added mention that the dataset profiling is not available through the Admin API

**Review by sid-acryl — APPROVED** [review]
LGTM

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**